packages/web3-eth-contract/src/contract.ts

```diff
@@ -735,22 +735,35 @@ export class Contract<Abi extends ContractAbi>
 
 		const filter = options?.filter ?? {};
 		const filterKeys = Object.keys(filter);
-		return eventName === 'allEvents' && filterKeys.length > 0
-			? decodedLogs.filter(log =>
-					typeof log === 'string'
-						? true
-						: filterKeys.every((k: string) =>
-								Array.isArray(filter[k])
-									? (filter[k] as Numbers[]).some(
-											(v: Numbers) =>
-												String(log.returnValues[k]).toUpperCase() ===
-												String(v).toUpperCase(),
-									  )
-									: String(log.returnValues[k]).toUpperCase() ===
-									  String(filter[k]).toUpperCase(),
-						  ),
-			  )
-			: decodedLogs;
+
+		if (filterKeys.length > 0) {
```

Comment:
This check was previously:

```typescript
eventName === 'allEvents' && filterKeys.length > 0
```

which was intended to short circuit the additional filter logic as it was expected that `logs` was already filtered by the event name by the RPC client. However, `logs` will be all events emitted during the transaction if `eventName === 'allEvents'` or we attempts to filter events using a non-indexed value as done in [this](https://github.com/web3/web3.js/blob/ef63c3237e9155dc7f7ff78af5f224ad087bd063/packages/web3-eth-contract/test/integration/contract_filter_events.test.ts#L245-L262) test